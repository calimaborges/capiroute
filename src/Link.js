import React from 'react';
import { router } from './';

const handleLink = (to, queryTo, onClick, keepQuery) => event => {
    event.preventDefault();
    if (onClick) {
        onClick(event);
    }

    if (to)
        router.goto(to, keepQuery);

    if (queryTo)
        router.setQueryString(queryTo);
};

const Link = ({ to, onClick, keepQuery, queryTo, children, ...props}) => (
    <a href={to || "/"}
        onClick={handleLink(to, queryTo, onClick, keepQuery)}
       {...props}>
        {children}
    </a>
);

Link.propTypes = {
    to: React.PropTypes.string,
    queryTo: React.PropTypes.object,
    keepQuery: React.PropTypes.bool
};

export default Link;