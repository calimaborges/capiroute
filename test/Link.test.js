import React from "react";
import Link from "../src/Link";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import { setCurrentUrl } from "./helper.lib";

test("should be rendered correctly", () => {
  const component = renderer.create(<Link to="/about">About</Link>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('should call goto if prop "to" is set', () => {
  const component = shallow(<Link to="/about">About</Link>);

  const oldPushState = window.history.pushState;
  const oldOnPopState = window.onpopstate;
  window.history.pushState = jest.fn();
  window.onpopstate = jest.fn();

  component.find("a").simulate("click", { preventDefault() {} });

  // FIXME: should mock router
  expect(window.history.pushState.mock.calls.length).toBe(1);
  expect(window.history.pushState.mock.calls[0][2]).toBe("/about");
  expect(window.onpopstate.mock.calls.length).toBe(1);

  window.history.pushState = oldPushState;
  window.onpopstate = oldOnPopState;
});

test("should call onClick if it is set", () => {
  let counter = 0;
  const component = shallow(<Link onClick={() => counter++}>About</Link>);

  component.find("a").simulate("click", { preventDefault() {} });

  expect(counter).toBe(1);
});

test('should set queryString if prop "queryTo" is set', () => {
  const component = shallow(<Link queryTo={{ type: "test" }}>About</Link>);

  setCurrentUrl("http://www.example.com/tasks");
  const oldPushState = window.history.pushState;
  const oldOnPopState = window.onpopstate;
  window.history.pushState = jest.fn();
  window.onpopstate = jest.fn();

  component.find("a").simulate("click", { preventDefault() {} });

  // FIXME: should mock router
  expect(window.history.pushState.mock.calls.length).toBe(1);
  expect(window.history.pushState.mock.calls[0][2]).toBe("/tasks?type=test");
  expect(window.onpopstate.mock.calls.length).toBe(1);

  window.history.pushState = oldPushState;
  window.onpopstate = oldOnPopState;
});

test('should set keep query if prop "keepQuery" is set', () => {
  const component = shallow(<Link to="/tasks" keepQuery>About</Link>);

  setCurrentUrl("http://www.example.com/?type=test");
  const oldPushState = window.history.pushState;
  const oldOnPopState = window.onpopstate;
  window.history.pushState = jest.fn();
  window.onpopstate = jest.fn();

  component.find("a").simulate("click", { preventDefault() {} });

  // FIXME: should mock router
  expect(window.history.pushState.mock.calls.length).toBe(1);
  expect(window.history.pushState.mock.calls[0][2]).toBe("/tasks?type=test");
  expect(window.onpopstate.mock.calls.length).toBe(1);

  window.history.pushState = oldPushState;
  window.onpopstate = oldOnPopState;
});
