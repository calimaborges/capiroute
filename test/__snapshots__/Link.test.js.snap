// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be rendered correctly 1`] = `
<a
  href="/about"
  onClick={[Function]}
>
  About
</a>
`;
